README.md
# elastic-query-export

Export Data from ElasticSearch to CSV by Raw or Lucene Query (e.g. from Kibana).
Works with ElasticSearch 6+ (OpenSearch works too) and makes use of ElasticSearch's Scroll API and Go's
concurrency possibilities to work as fast as possible.

## Requirements

Sync dependencies:

 ```shell
 go mod tidy
 ```

## Usage

````bash
es-query-export -c "http://localhost:9200" -i "logstash-*" --start="2019-04-04T12:15:00" --fields="RemoteHost,RequestTime,Timestamp,RequestUri,RequestProtocol,Agent" -q "RequestUri:*export*"
````

## CLI Options

| Flag                | Default               |                                                                                                         |
|---------------------|-----------------------|---------------------------------------------------------------------------------------------------------|
| `-h --help`         |                       | show help                                                                                               |
| `-v --version`      |                       | show version                                                                                            |
| `-c --connect`      | http://localhost:9200 | URI to ElasticSearch instance                                                                           |
| `-i --index`        | logs-*                | name of index to use, use globbing characters * to match multiple                                       |
| `-q --query`        |                       | Lucene query to match documents (same as in Kibana)                                                     |
| `   --fields`       |                       | define a comma separated list of fields to export                                                       |
| `-o --outpath`      | /opt/sixlabs/reports  | output path                                                                                             |
| `-fn --filename`    |                       | filename output the default name is de date in this format 20060102150405                               |
| `-f --outformat`    | csv                   | format of the output data: possible values csv, json, raw                                               |
| `-r --rawquery`     |                       | optional raw ElasticSearch query JSON string                                                            |
| `-s --start`        |                       | optional start date - Format: YYYY-MM-DDThh:mm:ss.SSSZ. or any other Elasticsearch default format       |
| `-e --end`          |                       | optional end date - Format: YYYY-MM-DDThh:mm:ss.SSSZ. or any other Elasticsearch default format         |
| `   --timefield`    |                       | optional time field to use, default to @timestamp                                                       |
| `   --verifySSL`    | true                  | optional define how to handle SSL certificates                                                          |
| `   --user`         |                       | optional username                                                                                       |
| `   --pass`         |                       | optional password                                                                                       |
| `   --size`         | 1000                  | size of the scroll window, the more the faster the export works but it adds more pressure on your nodes |
| `   --trace`        | false                 | enable trace mode to debug queries send to ElasticSearch                                                |
| `   --timezone`     | America/Santiago      | timezone to use for timestamps                                                                          |
| ` --convertdate`    | false                 | Convert UTC to date format with timezone in response                                                    |
| ` --dateformat`     | 2006-01-02 15:04:05   | Date format to use for convert dates to default or set format                                           |

## Output Formats

- `csv` - all or selected fields separated by comma (,) with field names in the first line 
- `json` - all or selected fields as JSON objects, one per line
- `raw` - JSON dump of matching documents including id, index and _source field containing the document data. One document as JSON object per line.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
### Added
 - [ ] add logging for develerpment purposes

## [1.1.0] - 2022-08-24
### Added
- Added Gitlab CI
- Added Dockerfile

## [1.0.0] - 2022-08-24
### Added
- Add timezone
- Add convert date with format
  

main.go
package main

import (
	"context"
	"fmt"
	"os"
	"os/signal"
	"syscall"

	"github.com/pteich/configstruct"

	"gitlab.com/sixbell/proyectos/brasil/claro/vas/elastic-query-export/export"
	"gitlab.com/sixbell/proyectos/brasil/claro/vas/elastic-query-export/flags"
)

var _ string

func main() {
	conf := flags.Flags{
		ElasticURL:       "http://localhost:9200",
		ElasticVerifySSL: true,
		Index:            "logs-*",
		Query:            "*",
		OutFormat:        flags.FormatCSV,
		Filename:         "",
		Outpath:          "/opt/sixlabs/reports",
		ScrollSize:       1000,
		Timefield:        "@timestamp",
		Convertdate:      false,
		Dateformat:       "2006-01-02 15:04:05",
		Timezone:         "America/Santiago",
		Debug:            false,
		LogLevel:         "info",
		LogOutput:        "/opt/sixlabs/logs",
		LogFormat:        "json",
		LogResult:        false,
		LogFile:          false,
	}

	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGKILL, syscall.SIGTERM, syscall.SIGINT)
	defer cancel()

	cmd := configstruct.NewCommand(
		"",
		"CLI tool to export data from ElasticSearch into a CSV or JSON file. https://gitlab.com/sixbell/proyectos/brasil/claro/vas/elastic-query-export",
		&conf,
		func(c *configstruct.Command, cfg interface{}) error {
			export.Run(ctx, cfg.(*flags.Flags))
			return nil
		},
	)

	err := cmd.ParseAndRun(os.Args)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	os.Exit(0)
}


formats/csv.go
package formats

import (
	"context"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"github.com/olivere/elastic/v7"
	"log"
	"os"
	"regexp"
	"sync"
	"time"

	"golang.org/x/sync/errgroup"

	"gitlab.com/sixbell/proyectos/brasil/claro/vas/elastic-query-export/flags"
)

type CSV struct {
	Conf    *flags.Flags
	Outfile *os.File
	Workers int
}

func (c CSV) Run(ctx context.Context, hits <-chan *elastic.SearchHit) error {
	g, ctx := errgroup.WithContext(ctx)

	csvout := make(chan []string, c.Workers)
	defer close(csvout)

	go func() {
		w := csv.NewWriter(c.Outfile)

		for csvdata := range csvout {
			if err := w.Write(csvdata); err != nil {
				log.Printf("Error writing CSV data - %v", err)
			}

			w.Flush()
		}

	}()

	sendHeader := sync.Once{}
	fields := c.Conf.Fields
	timezone := c.Conf.Timezone
	dateFormat := c.Conf.Dateformat

	headerSent := make(chan struct{})

	for i := 0; i < c.Workers; i++ {
		g.Go(func() error {
			var document map[string]interface{}

			for hit := range hits {
				var csvdata []string
				var outdata string

				if err := json.Unmarshal(hit.Source, &document); err != nil {
					log.Printf("Error unmarshal JSON from ElasticSearch - %v", err)
				}

				sendHeader.Do(func() {
					if c.Conf.Fields == nil {
						for key := range document {
							fields = append(fields, key)
						}
					}
					csvout <- fields
					close(headerSent)
				})

				<-headerSent

				for _, field := range fields {
					if val, ok := document[field]; ok {
						if val == nil {
							csvdata = append(csvdata, "")
							continue
						}

						// this type switch is probably not really needed anymore
						switch val := val.(type) {
						case int64:
							outdata = fmt.Sprintf("%d", val)
						case float64:
							d := int(val)
							if val == float64(d) {
								outdata = fmt.Sprintf("%d", d)
							} else {
								outdata = fmt.Sprintf("%f", val)
							}
						case []interface{}:
							outdata = fmt.Sprintf("%v", val)
						default:
							if val == nil {
								outdata = ""
							} else {
								outdata = fmt.Sprintf("%v", val)
							}

							if isDateValue(val.(string)) {
								if c.Conf.Convertdate {
									outdata = removeLBR(convertDate(val.(string), timezone, dateFormat))
								} else {
									outdata = removeLBR(fmt.Sprintf("%s", formatDate(val.(string), dateFormat)))
								}
							} else {
								outdata = removeLBR(fmt.Sprintf("%s", val))
							}
						}

						csvdata = append(csvdata, outdata)
					} else {
						csvdata = append(csvdata, "")
					}
				}

				// send string array to csv output
				csvout <- csvdata

				select {
				default:
				case <-ctx.Done():
					return ctx.Err()
				}
			}
			return nil
		})
	}

	return g.Wait()
}

func removeLBR(text string) string {
	re := regexp.MustCompile(`\x{000D}\x{000A}|[\x{000A}\x{000B}\x{000C}\x{000D}\x{0085}\x{2028}\x{2029}]`)
	return re.ReplaceAllString(text, ``)
}

func isDateValue(stringDate string) bool {
	timeLayout := "2006-01-02T15:04:05.000Z"
	_, err := time.Parse(timeLayout, stringDate)
	return err == nil
}

func convertDate(stringDate string, timeZone string, dateformat string) string {
	timeLayout := "2006-01-02T15:04:05.000Z"
	locTime, err := time.LoadLocation(timeZone)
	if err != nil {
		fmt.Println(err)
	}
	timeParam, err := time.Parse(timeLayout, stringDate)
	if err != nil {
		fmt.Println(err)
	}

	return fmt.Sprintf("%v", timeParam.In(locTime).Format(dateformat))
}

func formatDate(stringDate string, dateformat string) string {
	timeLayout := "2006-01-02T15:04:05.000Z"
	timeParam, err := time.Parse(timeLayout, stringDate)
	if err != nil {
		fmt.Println(err)
	}
	return fmt.Sprintf("%v", timeParam.Format(dateformat))
}


export/export.go
package export

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"github.com/olivere/elastic/v7"
	"gitlab.com/sixbell/proyectos/brasil/claro/vas/elastic-query-export/logger"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"

	"gitlab.com/sixbell/proyectos/brasil/claro/vas/elastic-query-export/flags"
	"gitlab.com/sixbell/proyectos/brasil/claro/vas/elastic-query-export/formats"
	logWrapper "gitlab.com/sixbell/proyectos/brasil/claro/vas/elastic-query-export/logger"
)

const workers = 8

// Formatter defines how an output formatter has to look like
type Formatter interface {
	Run(context.Context, <-chan *elastic.SearchHit) error
}

type RawQuery struct {
	Query *Query `json:"query,omitempty"`
}

type Query struct {
	Bool *Bool `json:"bool,omitempty"`
}

type Bool struct {
	Filter *Filter `json:"filter,omitempty"`
	Must   *Must   `json:"must,omitempty"`
}

type Filter struct {
	Range *Range `json:"range,omitempty"`
}

type Range struct {
	Timestamp *Timestamp `json:"@timestamp,omitempty"`
}

type Timestamp struct {
	From         *string `json:"from,omitempty"`
	IncludeLower *bool   `json:"include_lower,omitempty"`
	IncludeUpper *bool   `json:"include_upper,omitempty"`
	To           *string `json:"to,omitempty"`
}

type Must struct {
	QueryString *QueryString `json:"query_string,omitempty"`
}

type QueryString struct {
	Query *string `json:"query,omitempty"`
}

// Run starts the export of Elastic data
func Run(ctx context.Context, conf *flags.Flags) {
	var loggerWrapper *logWrapper.StandardLogger

	if conf.Debug {
		loggerWrapper = logger.NewLogger(conf)
	}

	if loggerWrapper != nil {
		loggerWrapper.Info("Starting export")
	}

	exportCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	tr := &http.Transport{
		TLSClientConfig: &tls.Config{InsecureSkipVerify: !conf.ElasticVerifySSL},
	}
	httpClient := &http.Client{Transport: tr}

	esOpts := make([]elastic.ClientOptionFunc, 0)
	esOpts = append(esOpts,
		elastic.SetHttpClient(httpClient),
		elastic.SetURL(conf.ElasticURL),
		elastic.SetSniff(false),
		elastic.SetHealthcheckInterval(60*time.Second),
		elastic.SetErrorLog(log.New(os.Stderr, "ELASTIC ", log.LstdFlags)),
	)

	if conf.Debug {
		esOpts = append(esOpts, elastic.SetTraceLog(log.New(os.Stderr, "ELASTIC ", log.LstdFlags)))
	}

	if conf.ElasticUser != "" && conf.ElasticPass != "" {
		esOpts = append(esOpts, elastic.SetBasicAuth(conf.ElasticUser, conf.ElasticPass))
	}

	client, err := elastic.NewClient(esOpts...)
	if err != nil {
		log.Fatalf("Error connecting to ElasticSearch: %s", err)
	}
	defer client.Stop()

	if conf.Fieldlist != "" {
		conf.Fields = strings.Split(conf.Fieldlist, ",")
	}

	outfile, err := os.Create(fullFilePath(conf.Outpath, conf.OutFormat, conf.Filename))
	if err != nil {
		log.Fatalf("Error creating output file - %s", err)
	}
	defer outfile.Close()

	var rangeQuery *elastic.RangeQuery

	esQuery := elastic.NewBoolQuery()

	if conf.StartDate != "" && conf.EndDate != "" {
		rangeQuery = elastic.NewRangeQuery(conf.Timefield).Gte(conf.StartDate).Lte(conf.EndDate)
	} else if conf.StartDate != "" {
		rangeQuery = elastic.NewRangeQuery(conf.Timefield).Gte(conf.StartDate)
	} else if conf.EndDate != "" {
		rangeQuery = elastic.NewRangeQuery(conf.Timefield).Lte(conf.EndDate)
	} else {
		rangeQuery = nil
	}

	if rangeQuery != nil {
		esQuery = esQuery.Filter(rangeQuery)
	}

	if conf.RAWQuery != "" {
		if loggerWrapper != nil {
			loggerWrapper.Debug(fmt.Sprintf("rawQuery: %s", conf.RAWQuery))
		}
		esQuery = esQuery.Must(elastic.NewRawStringQuery(conf.RAWQuery))
	} else if conf.Query != "" {
		if loggerWrapper != nil {
			loggerWrapper.Debug(fmt.Sprintf("query: %s", conf.Query))
		}

		newQuery, err := parseQuery(conf.Query, loggerWrapper)
		if err != nil {
			if loggerWrapper != nil {
				loggerWrapper.Debug(fmt.Sprintf("Error parsing query: %s", err))
			}
		}
		esQuery = esQuery.Must(elastic.NewQueryStringQuery(newQuery))
	} else {
		esQuery = esQuery.Must(elastic.NewMatchAllQuery())
	}

	// Count total and setup progress
	total, err := client.Count(conf.Index).Query(esQuery).Do(ctx)
	if err != nil {
		if loggerWrapper != nil {
			loggerWrapper.Error(fmt.Sprintf("Error counting ElasticSearch documents - %v", err))
		}
	}

	if conf.LogResult {
		if loggerWrapper != nil {
			loggerWrapper.Debug(fmt.Sprintf("Total documents to export: %d", total))
		}
	}

	hits := make(chan *elastic.SearchHit)

	go func() {
		defer close(hits)

		scroll := client.Scroll(conf.Index).Size(conf.ScrollSize).Query(esQuery)

		// include selected fields otherwise export all
		if conf.Fields != nil {
			fetchSource := elastic.NewFetchSourceContext(true)
			for _, field := range conf.Fields {
				fetchSource.Include(field)
			}
			scroll = scroll.FetchSourceContext(fetchSource)
		}

		if conf.LogResult {
			if loggerWrapper != nil {
				data, _ := json.MarshalIndent(scroll.Human(true), "", "  ")
				loggerWrapper.Debug(fmt.Sprintf("ElasticSearch query: %s", string(data)))
			}
		}

		for {
			results, err := scroll.Do(ctx)
			if err == io.EOF {
				return // all results retrieved
			}
			if err != nil {
				if loggerWrapper != nil {
					loggerWrapper.Error(fmt.Sprintf("Error scrolling ElasticSearch documents - %v", err))
				}
				cancel()
				return // something went wrong
			}

			// Send the hits to the hits channel
			for _, hit := range results.Hits.Hits {
				// Check if we need to terminate early
				select {
				case hits <- hit:
				case <-exportCtx.Done():
					return
				}
			}
		}
	}()

	var output Formatter
	switch conf.OutFormat {
	case flags.FormatJSON:
		output = formats.JSON{
			Outfile: outfile,
		}
	case flags.FormatRAW:
		output = formats.Raw{
			Outfile: outfile,
		}
	default:
		output = formats.CSV{
			Conf:    conf,
			Outfile: outfile,
			Workers: workers,
		}
	}

	err = output.Run(exportCtx, hits)
	if err != nil {
		if loggerWrapper != nil {
			loggerWrapper.Error(fmt.Sprintf("Error exporting ElasticSearch documents - %v", err))
		}
	}
}

func fullFilePath(path string, format string, filename string) string {
	if filename != "" {
		return path + "/" + filename + "." + format
	}
	return path + "/" + time.Now().Format("20060102150405") + "." + format
}

func parseQuery(query string, loggerWrapper *logWrapper.StandardLogger) (string, error) {
	strs := strings.Split(query, " ")
	var newQuery []string

	for i, str := range strs {
		if loggerWrapper != nil {
			loggerWrapper.Debug(fmt.Sprintf("str: %s %d", str, i))
		}
		if strings.Count(str, ":") > 1 {
			countChar := strings.Count(str, ":")
			var tempString = str
			for j := 1; j < countChar; j++ {
				tempString = replaceNth(tempString, ":", "\\:", j+1)
			}
			newQuery = append(newQuery, tempString)
			continue
		}

		newQuery = append(newQuery, str)
	}

	query = strings.Join(newQuery, " ")
	if loggerWrapper != nil {
		loggerWrapper.Debug(fmt.Sprintf("parse query: %s", query))
	}

	return query, nil
}

func replaceNth(s, old, new string, n int) string {
	i := 0
	for m := 1; m <= n; m++ {
		x := strings.Index(s[i:], old)
		if x < 0 {
			break
		}
		i += x
		if m == n {
			return s[:i] + new + s[i+len(old):]
		}
		i += len(old)
	}
	return s
}


logger/logger.go
package logger

import (
	"github.com/sirupsen/logrus"
	"gitlab.com/sixbell/proyectos/brasil/claro/vas/elastic-query-export/flags"
	"io"
	"log"
	"os"
)

// Event stores messages to log later, from our standard interface
type Event struct {
	id      int
	message string
}

// StandardLogger enforces specific log message formats
type StandardLogger struct {
	*logrus.Logger
}

var confFlags *flags.Flags

// NewLogger initializes the standard logger
func NewLogger(conf *flags.Flags) *StandardLogger {
	var file *os.File
	var baseLogger = logrus.New()
	confFlags = conf

	var standardLogger = &StandardLogger{baseLogger}

	if conf.LogFile {
		standardLogger.SetOutput(file)
	}

	switch conf.LogLevel {
	case "debug":
		standardLogger.SetLevel(logrus.DebugLevel)
	case "info":
		standardLogger.SetLevel(logrus.InfoLevel)
	case "warn":
		standardLogger.SetLevel(logrus.WarnLevel)
	case "error":
		standardLogger.SetLevel(logrus.ErrorLevel)
	case "fatal":
		standardLogger.SetLevel(logrus.FatalLevel)
	case "panic":
		standardLogger.SetLevel(logrus.PanicLevel)
	default:
		standardLogger.SetLevel(logrus.InfoLevel)
	}

	if conf.LogFormat == "json" {
		standardLogger.Formatter = &logrus.JSONFormatter{}
	} else {
		standardLogger.Formatter = &logrus.TextFormatter{}
	}

	if conf.LogFile {
		closeFile(file)
	}
	return standardLogger
}

// Declare variables to store log messages as new Events
var (
	invalidArgMessage      = Event{1, "Invalid arg: %s"}
	invalidArgValueMessage = Event{2, "Invalid value for argument: %s: %v"}
	missingArgMessage      = Event{3, "Missing arg: %s"}
	infoMessage            = Event{4, " %s"}
	debugMessage           = Event{5, " %s"}
)

// InvalidArg is a standard error message
func (l *StandardLogger) InvalidArg(argumentName string) {
	logToFile(l)
	l.Errorf(invalidArgMessage.message, argumentName)
}

// InvalidArgValue is a standard error message
func (l *StandardLogger) InvalidArgValue(argumentName string, argumentValue string) {
	logToFile(l)
	l.Errorf(invalidArgValueMessage.message, argumentName, argumentValue)
}

// MissingArg is a standard error message
func (l *StandardLogger) MissingArg(argumentName string) {
	logToFile(l)
	l.Errorf(missingArgMessage.message, argumentName)
}

// Info is a standard info message
func (l *StandardLogger) Info(message string) {
	logToFile(l)
	l.Infof(infoMessage.message, message)
}

// Debug is a standard debug message
func (l *StandardLogger) Debug(message string) {
	logToFile(l)
	l.Debugf(debugMessage.message, message)
}

// Log is a standard log message
func (l *StandardLogger) Log(message string) {
	logToFile(l)
	l.Log(message)
}

func openFile() *os.File {
	file, err := os.OpenFile("elastic-query-export.log", os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		log.Fatal(err)
	}

	return file
}

func closeFile(file *os.File) {
	err := file.Close()
	if err != nil {
		return
	}
}

func logToFile(l *StandardLogger) {
	if confFlags.LogFile {
		var file = openFile()
		mw := io.MultiWriter(os.Stdout, file)
		l.SetOutput(mw)
		closeFile(file)
	}
}


flags/flags.go
package flags

const (
	FormatCSV  = "csv"
	FormatJSON = "json"
	FormatRAW  = "raw"
)

type Flags struct {
	ElasticURL       string `cli:"connect" cliAlt:"c" usage:"ElasticSearch URL"`
	ElasticUser      string `cli:"user" usage:"ElasticSearch Username"`
	ElasticPass      string `cli:"pass" usage:"ElasticSearch Password"`
	ElasticVerifySSL bool   `cli:"verifySSL" usage:"Verify SSL certificate"`
	Index            string `cli:"index" cliAlt:"i" usage:"ElasticSearch Index (or Index Prefix)"`
	RAWQuery         string `cli:"rawquery" cliAlt:"r" usage:"ElasticSearch raw query string"`
	Query            string `cli:"query" cliAlt:"q" usage:"Lucene query same that is used in Kibana search input"`
	OutFormat        string `cli:"outformat" cliAlt:"f" usage:"Format of the output data. [json|csv]"`
	Filename         string `cli:"filename" cliAlt:"fn" usage:"File name output"`
	Outpath          string `cli:"outpath" cliAlt:"o" usage:"Path to output file"`
	StartDate        string `cli:"start" cliAlt:"s" usage:"Start date for included documents"`
	EndDate          string `cli:"end" cliAlt:"e" usage:"End date for included documents"`
	ScrollSize       int    `cli:"size" usage:"Number of documents that will be returned per shard"`
	Timefield        string `cli:"timefield" usage:"Field name to use for start and end date query"`
	Timezone         string `cli:"timezone" usage:"Timezone to use for parse dates response"`
	Fieldlist        string `cli:"fields" usage:"Fields to include in export as comma separated list"`
	Convertdate      bool   `cli:"convertdate" usage:"Convert UTC to date format with timezone in response"`
	Dateformat       string `cli:"dateformat" usage:"Date format to use for convert dates default 2006-01-12 15:03:05"`
	Trace            bool   `cli:"trace" usage:"Enable debug output"`
	Debug            bool   `cli:"logconsole" usage:"Enable debug output"`
	LogLevel         string `cli:"loglevel" usage:"Log level"`
	LogOutput        string `cli:"logoutput" usage:"Log output path"`
	LogFile          bool   `cli:"logfile" usage:"Log to file"`
	LogFormat        string `cli:"logformat" usage:"Log format [text|json]"`
	LogResult        bool   `cli:"logresult" usage:"Log elasticsearch query result"`
	Fields           []string
}
